### ISSUE TITLE
Worlds and Entities can have render() called before first update()

### ISSUE THREAD

**Issue by Draknek** [useflashpunk/FlashPunk#99]
Because checkWorld() is called after world.update() not before, new worlds are rendered for the first time before they're updated for the first time.

Similarly, updateLists() is called after world.update(), so newly added entities are rendered before they're first updated.

This reduces render lag by one frame but may cause graphical weirdness in some games.

**Comment by Draknek** [comment]
Fixed in 93dd26d052cc11fd6891ded31f6d8481b00c1f41 (hopefully without other effects)

**Comment by Draknek** [comment]
Discussion in 93dd26d052cc11fd6891ded31f6d8481b00c1f41:
- Changing an entity's position and removing it in the same frame makes it appear for one frame at the new position.
- Expectation: once an entity is removed, it should no longer be rendered.

My proposed change: both add and remove at the start of the frame, and remove but not add between update and render.

**Comment by danishgoel** [comment]
Though your suggestion solves the immediate issue, but it may cause more confusion. As entities will now be removed in the same frame but added in the next one. If I would try to swap out an entity it would still flicker as there would be now no entity to render in the current frame.

Only real solution (as far as I can see) would be with a mechanism that does both add and remove at the same point in the loop.

Maybe we can have an optional argument to both add and remove functions that specifies whether to wait till frame end to add/remove, or add/remove immediately. e.g.

``` as3
public function add(e:Entity, immediate:Boolean = false):Entity { ... }

public function remove(e:Entity, immediate:Boolean = false):Entity { ... }
```

this way we can use the optional argument in the special cases where we want the change to be effective immediately. Or we can have it the other way around and use the old method by default.

**Comment by kylepulver** [comment]
When changing the layer of an entity during its update you can get some pretty wacky results.  For example I have an entity on layer -100, and an entity on layer 0.  If I set the entity on layer 0 to -200, the entity on layer -100 will disappear for a frame.  If I switch that -200 entity back to 0, the entity itself will disappear for a frame.

**Comment by Draknek** [comment]
Fixed in b1708223a0a56c875cf2f89d559ce436dc3f2627.

@danishgoel @kylepulver Can you check this out and see if it fits your expected behaviour now?
